File: src/platform_context_graph/data_intelligence/dbt_sql_lineage.py
"""Compiled SQL lineage helpers for dbt-style manifest normalization."""

from __future__ import annotations

import re
from collections.abc import Mapping, Sequence
from dataclasses import dataclass

from .dbt_sql_expressions import (
    derived_expression_gap,
    expression_ignored_identifiers,
    expression_partial_reason,
)
from .dbt_sql_identifiers import unqualified_identifiers

_SELECT_CLAUSE_RE = re.compile(
    r"\bselect\b(?P<select>.*?)\bfrom\b", re.IGNORECASE | re.DOTALL
)
_AS_ALIAS_RE = re.compile(
    r"^(?P<expression>.+?)\s+as\s+(?P<alias>[A-Za-z_][A-Za-z0-9_]*)$",
    re.IGNORECASE | re.DOTALL,
)
_QUALIFIED_REFERENCE_RE = re.compile(
    r"\b(?P<alias>[A-Za-z_][A-Za-z0-9_]*)\."
    r"(?P<column>\*|[A-Za-z_][A-Za-z0-9_]*)(?=[^A-Za-z0-9_]|$)"
)
_FROM_RELATION_RE = re.compile(
    r"\b(?:from|join|left\s+join|right\s+join|inner\s+join|full\s+join|cross\s+join)"
    r"\s+(?P<relation>[A-Za-z_][A-Za-z0-9_]*(?:\.[A-Za-z_][A-Za-z0-9_]*){0,2})"
    r"(?:\s+(?:as\s+)?(?P<alias>[A-Za-z_][A-Za-z0-9_]*))?",
    re.IGNORECASE,
)


@dataclass(frozen=True, slots=True)
class ColumnLineage:
    """Normalized lineage for one output column."""

    output_column: str
    source_columns: tuple[str, ...]


@dataclass(frozen=True, slots=True)
class CompiledModelLineage:
    """Normalized lineage extracted from one compiled SQL model."""

    column_lineage: tuple[ColumnLineage, ...]
    unresolved_references: tuple[dict[str, str], ...]
    projection_count: int


@dataclass(frozen=True, slots=True)
class _RelationBinding:
    """Resolved relation metadata for one source alias in a SQL statement."""

    asset_name: str | None
    column_names: tuple[str, ...]
    column_lineage: Mapping[str, tuple[str, ...]]


def extract_compiled_model_lineage(
    compiled_sql: str,
    *,
    model_name: str,
    relation_column_names: Mapping[str, Sequence[str]],
) -> CompiledModelLineage:
    """Extract final model lineage, including simple CTE propagation."""

    cte_queries, final_sql = _split_cte_queries(compiled_sql)
    cte_bindings: dict[str, _RelationBinding] = {}
    unresolved_references: list[dict[str, str]] = []

    for cte_name, cte_sql in cte_queries:
        cte_lineage = _extract_select_lineage(
            cte_sql,
            model_name=model_name,
            relation_bindings=_relation_bindings(
                cte_sql,
                relation_column_names=relation_column_names,
                cte_bindings=cte_bindings,
            ),
        )
        unresolved_references.extend(cte_lineage.unresolved_references)
        cte_bindings[cte_name] = _binding_for_column_lineage(cte_lineage.column_lineage)

    final_lineage = _extract_select_lineage(
        final_sql,
        model_name=model_name,
        relation_bindings=_relation_bindings(
            final_sql,
            relation_column_names=relation_column_names,
            cte_bindings=cte_bindings,
        ),
    )
    unresolved_references.extend(final_lineage.unresolved_references)
    return CompiledModelLineage(
        column_lineage=final_lineage.column_lineage,
        unresolved_references=tuple(unresolved_references),
        projection_count=final_lineage.projection_count,
    )


def _binding_for_column_lineage(
    column_lineage: Sequence[ColumnLineage],
) -> _RelationBinding:
    """Build a CTE relation binding from projected column lineage."""

    lineage_by_name: dict[str, tuple[str, ...]] = {}
    column_names: list[str] = []
    for item in column_lineage:
        lineage_by_name[item.output_column] = item.source_columns
        column_names.append(item.output_column)
    return _RelationBinding(
        asset_name=None,
        column_names=tuple(column_names),
        column_lineage=lineage_by_name,
    )


def _relation_bindings(
    sql: str,
    *,
    relation_column_names: Mapping[str, Sequence[str]],
    cte_bindings: Mapping[str, _RelationBinding],
) -> dict[str, _RelationBinding]:
    """Return relation bindings for asset and CTE aliases in one SQL statement."""

    bindings: dict[str, _RelationBinding] = {}
    for match in _FROM_RELATION_RE.finditer(sql):
        relation_name = match.group("relation").strip()
        alias = (match.group("alias") or "").strip()
        relation_binding = cte_bindings.get(relation_name)
        if relation_binding is None:
            relation_binding = _RelationBinding(
                asset_name=relation_name,
                column_names=tuple(relation_column_names.get(relation_name, ())),
                column_lineage={},
            )

        names = {relation_name}
        if alias and alias.lower() not in {"on", "where", "group", "order", "limit"}:
            names.add(alias)
        else:
            names.add(relation_name.rsplit(".", maxsplit=1)[-1])

        for binding_name in names:
            bindings[binding_name] = relation_binding
    return bindings


def _extract_select_lineage(
    sql: str,
    *,
    model_name: str,
    relation_bindings: Mapping[str, _RelationBinding],
) -> CompiledModelLineage:
    """Extract lineage from one concrete SELECT statement."""

    column_lineage: list[ColumnLineage] = []
    unresolved_references: list[dict[str, str]] = []
    select_items = _extract_select_items(sql)

    for select_item in select_items:
        projection = _lineage_for_projection(
            select_item=select_item,
            relation_bindings=relation_bindings,
            model_name=model_name,
        )
        column_lineage.extend(projection.column_lineage)
        unresolved_references.extend(projection.unresolved_references)

    return CompiledModelLineage(
        column_lineage=tuple(column_lineage),
        unresolved_references=tuple(unresolved_references),
        projection_count=len(select_items),
    )


def _split_cte_queries(compiled_sql: str) -> tuple[list[tuple[str, str]], str]:
    """Split a SQL string into ordered CTE queries and the final SELECT."""

    trimmed = compiled_sql.lstrip()
    if not trimmed[:4].lower() == "with":
        return [], compiled_sql

    queries: list[tuple[str, str]] = []
    index = 4
    length = len(trimmed)

    while index < length:
        while index < length and trimmed[index].isspace():
            index += 1
        name_start = index
        while index < length and (
            trimmed[index].isalnum() or trimmed[index] == "_"
        ):
            index += 1
        cte_name = trimmed[name_start:index].strip()
        if not cte_name:
            break

        while index < length and trimmed[index].isspace():
            index += 1
        if index < length and trimmed[index] == "(":
            index = _consume_balanced_segment(trimmed, index)
            while index < length and trimmed[index].isspace():
                index += 1

        if trimmed[index : index + 2].lower() != "as":
            break
        index += 2
        while index < length and trimmed[index].isspace():
            index += 1
        if index >= length or trimmed[index] != "(":
            break

        query_start = index + 1
        query_end = _consume_balanced_segment(trimmed, index) - 1
        queries.append((cte_name, trimmed[query_start:query_end]))
        index = query_end + 1
        while index < length and trimmed[index].isspace():
            index += 1
        if index < length and trimmed[index] == ",":
            index += 1
            continue
        return queries, trimmed[index:]

    return [], compiled_sql


def _consume_balanced_segment(text: str, start_index: int) -> int:
    """Return the index immediately after one balanced parenthesized segment."""

    depth = 0
    index = start_index
    while index < len(text):
        character = text[index]
        if character == "(":
            depth += 1
        elif character == ")":
            depth -= 1
            if depth == 0:
                return index + 1
        index += 1
    return len(text)


def _extract_select_items(compiled_sql: str) -> list[str]:
    """Extract top-level SELECT projections from SQL text."""

    match = _SELECT_CLAUSE_RE.search(compiled_sql)
    if match is None:
        return []

    select_clause = match.group("select")
    items: list[str] = []
    current: list[str] = []
    depth = 0
    in_single_quote = False

    for character in select_clause:
        if character == "'" and (not current or current[-1] != "\\"):
            in_single_quote = not in_single_quote
        elif not in_single_quote:
            if character == "(":
                depth += 1
            elif character == ")" and depth > 0:
                depth -= 1
            elif character == "," and depth == 0:
                item = "".join(current).strip()
                if item:
                    items.append(item)
                current = []
                continue
        current.append(character)

    tail = "".join(current).strip()
    if tail:
        items.append(tail)
    return items


def _lineage_for_projection(
    *,
    select_item: str,
    relation_bindings: Mapping[str, _RelationBinding],
    model_name: str,
) -> CompiledModelLineage:
    """Extract supported source-column lineage from one SELECT item."""

    match = _AS_ALIAS_RE.match(select_item.strip())
    if match:
        expression = match.group("expression").strip()
        output_column = match.group("alias").strip()
    else:
        expression = select_item.strip()
        output_column = _implicit_output_column(expression)

    column_lineage: list[ColumnLineage] = []
    source_columns: list[str] = []
    unresolved_references: list[dict[str, str]] = []
    seen_columns: set[str] = set()
    matched_unqualified_identifiers: set[str] = set()
    matched_unqualified_identifiers.update(expression_ignored_identifiers(expression))

    for reference in _QUALIFIED_REFERENCE_RE.finditer(expression):
        alias = reference.group("alias")
        column = reference.group("column")
        binding = relation_bindings.get(alias)
        matched_unqualified_identifiers.add(alias)
        matched_unqualified_identifiers.add(column)

        resolved_columns = _resolve_reference_columns(
            binding,
            alias=alias,
            column=column,
            model_name=model_name,
        )
        if isinstance(resolved_columns, dict):
            unresolved_references.append(resolved_columns)
            continue
        if column == "*":
            for expanded_output_column, expanded_source_columns in resolved_columns:
                column_lineage.append(
                    ColumnLineage(
                        output_column=expanded_output_column,
                        source_columns=expanded_source_columns,
                    )
                )
            continue
        for source_column in resolved_columns:
            if source_column in seen_columns:
                continue
            seen_columns.add(source_column)
            source_columns.append(source_column)

    for identifier in unqualified_identifiers(
        expression,
        matched_identifiers=matched_unqualified_identifiers,
    ):
        resolved_columns = _resolve_unqualified_reference_columns(
            identifier,
            relation_bindings=relation_bindings,
            model_name=model_name,
        )
        if isinstance(resolved_columns, dict):
            unresolved_references.append(resolved_columns)
            continue
        for source_column in resolved_columns:
            if source_column in seen_columns:
                continue
            seen_columns.add(source_column)
            source_columns.append(source_column)

    if output_column is None and source_columns:
        output_column = source_columns[0].rsplit(".", maxsplit=1)[-1]
    partial_reason = expression_partial_reason(expression)
    if source_columns and partial_reason is not None:
        unresolved_references.append(
            derived_expression_gap(
                expression=expression,
                model_name=model_name,
                reason=partial_reason,
            )
        )
    if output_column is not None and source_columns:
        column_lineage.append(
            ColumnLineage(
                output_column=output_column.strip(),
                source_columns=tuple(source_columns),
            )
        )

    return CompiledModelLineage(
        column_lineage=tuple(column_lineage),
        unresolved_references=tuple(unresolved_references),
        projection_count=1,
    )


def _resolve_reference_columns(
    binding: _RelationBinding | None,
    *,
    alias: str,
    column: str,
    model_name: str,
) -> Sequence[str] | Sequence[tuple[str, tuple[str, ...]]] | dict[str, str]:
    """Resolve one qualified column reference through an asset or CTE binding."""

    if binding is None:
        return {
            "expression": f"{alias}.{column}",
            "model_name": model_name,
            "reason": "source_alias_not_resolved",
        }
    if column == "*":
        if binding.asset_name is not None:
            if not binding.column_names:
                return {
                    "expression": f"{alias}.*",
                    "model_name": model_name,
                    "reason": "wildcard_projection_not_supported",
                }
            return [
                (
                    expanded_column,
                    (f"{binding.asset_name}.{expanded_column}",),
                )
                for expanded_column in binding.column_names
            ]
        if not binding.column_names:
            return {
                "expression": f"{alias}.*",
                "model_name": model_name,
                "reason": "wildcard_projection_not_supported",
            }
        return [
            (
                expanded_column,
                binding.column_lineage.get(expanded_column, ()),
            )
            for expanded_column in binding.column_names
            if binding.column_lineage.get(expanded_column)
        ]

    if binding.asset_name is not None:
        return (f"{binding.asset_name}.{column}",)

    source_columns = tuple(binding.column_lineage.get(column, ()))
    if not source_columns:
        return {
            "expression": f"{alias}.{column}",
            "model_name": model_name,
            "reason": "cte_column_not_resolved",
        }
    return source_columns


def _resolve_unqualified_reference_columns(
    identifier: str,
    *,
    relation_bindings: Mapping[str, _RelationBinding],
    model_name: str,
) -> tuple[str, ...] | dict[str, str]:
    """Resolve one bare identifier when it maps to a unique visible source."""

    candidates: dict[tuple[str, ...], tuple[str, ...]] = {}
    for binding in relation_bindings.values():
        resolved_columns = _binding_columns_for_identifier(binding, identifier)
        if resolved_columns:
            candidates.setdefault(tuple(resolved_columns), tuple(resolved_columns))

    if not candidates:
        return {
            "expression": identifier,
            "model_name": model_name,
            "reason": "unqualified_column_reference_not_resolved",
        }
    if len(candidates) > 1:
        return {
            "expression": identifier,
            "model_name": model_name,
            "reason": "unqualified_column_reference_ambiguous",
        }
    return next(iter(candidates.values()))


def _binding_columns_for_identifier(
    binding: _RelationBinding,
    identifier: str,
) -> tuple[str, ...]:
    """Return resolved source columns for one identifier within one binding."""

    if binding.asset_name is not None:
        if identifier not in binding.column_names:
            return ()
        return (f"{binding.asset_name}.{identifier}",)
    return tuple(binding.column_lineage.get(identifier, ()))


def _implicit_output_column(expression: str) -> str | None:
    """Infer an output column name when a projection omits `AS alias`."""

    references = list(_QUALIFIED_REFERENCE_RE.finditer(expression))
    if len(references) == 1 and references[0].group("column") != "*":
        return references[0].group("column")
    return None


__all__ = ["ColumnLineage", "CompiledModelLineage", "extract_compiled_model_lineage"]

File: go/internal/parser/dbt_sql_lineage.go
package parser

import (
	"regexp"
	"strings"
)

// ColumnLineage describes one output column and the source columns that feed it.
type ColumnLineage struct {
	OutputColumn        string
	SourceColumns       []string
	TransformKind       string
	TransformExpression string
}

// CompiledModelLineage summarizes lineage extracted from one compiled dbt model.
type CompiledModelLineage struct {
	ColumnLineage        []ColumnLineage
	UnresolvedReferences []map[string]string
	ProjectionCount      int
}

type relationBinding struct {
	assetName     string
	columnNames   []string
	columnLineage map[string]ColumnLineage
}

var (
	dbtSelectClauseRe = regexp.MustCompile(`(?is)\bselect\b(?P<select>.*?)\bfrom\b`)
	dbtAsAliasRe      = regexp.MustCompile(`(?is)^(?P<expression>.+?)\s+as\s+(?P<alias>[A-Za-z_][A-Za-z0-9_]*)$`)
	dbtFromRelationRe = regexp.MustCompile(`(?i)\b(?:from|join|left\s+join|right\s+join|inner\s+join|full\s+join|cross\s+join)\s+(?P<relation>[A-Za-z_][A-Za-z0-9_]*(?:\.[A-Za-z_][A-Za-z0-9_]*){0,2})(?:\s+(?:as\s+)?(?P<alias>[A-Za-z_][A-Za-z0-9_]*))?`)
)

func extractCompiledModelLineage(compiledSQL string, modelName string, relationColumnNames map[string][]string) CompiledModelLineage {
	cteQueries, finalSQL := splitCTEQueries(compiledSQL)
	cteBindings := make(map[string]*relationBinding)
	unresolved := make([]map[string]string, 0)

	for _, cteQuery := range cteQueries {
		cteName := cteQuery[0]
		cteSQL := cteQuery[1]
		cteLineage := extractSelectLineage(cteSQL, modelName, relationBindings(cteSQL, relationColumnNames, cteBindings))
		unresolved = append(unresolved, cteLineage.UnresolvedReferences...)
		cteBindings[cteName] = bindingForColumnLineage(cteLineage.ColumnLineage)
	}

	finalLineage := extractSelectLineage(finalSQL, modelName, relationBindings(finalSQL, relationColumnNames, cteBindings))
	unresolved = append(unresolved, finalLineage.UnresolvedReferences...)
	return CompiledModelLineage{
		ColumnLineage:        finalLineage.ColumnLineage,
		UnresolvedReferences: unresolved,
		ProjectionCount:      finalLineage.ProjectionCount,
	}
}

func bindingForColumnLineage(lineage []ColumnLineage) *relationBinding {
	lineageByName := make(map[string]ColumnLineage)
	columnNames := make([]string, 0, len(lineage))
	for _, item := range lineage {
		lineageByName[item.OutputColumn] = item
		columnNames = append(columnNames, item.OutputColumn)
	}
	return &relationBinding{
		columnNames:   columnNames,
		columnLineage: lineageByName,
	}
}

func relationBindings(sql string, relationColumnNames map[string][]string, cteBindings map[string]*relationBinding) map[string]*relationBinding {
	bindings := make(map[string]*relationBinding)
	for _, match := range dbtFromRelationRe.FindAllStringSubmatch(sql, -1) {
		relationName := strings.TrimSpace(match[1])
		alias := strings.TrimSpace(match[2])
		binding := cteBindings[relationName]
		if binding == nil {
			binding = &relationBinding{
				assetName:     relationName,
				columnNames:   relationColumnNames[relationName],
				columnLineage: map[string]ColumnLineage{},
			}
		}
		names := map[string]struct{}{relationName: {}}
		if alias != "" && !isReservedAlias(alias) {
			names[alias] = struct{}{}
		} else if parts := strings.Split(relationName, "."); len(parts) > 0 {
			names[parts[len(parts)-1]] = struct{}{}
		}
		for name := range names {
			bindings[name] = binding
		}
	}
	return bindings
}

func extractSelectLineage(sql string, modelName string, bindings map[string]*relationBinding) CompiledModelLineage {
	columnLineage := make([]ColumnLineage, 0)
	unresolved := make([]map[string]string, 0)
	selectItems := extractSelectItems(sql)
	for _, selectItem := range selectItems {
		projection := lineageForProjection(selectItem, bindings, modelName)
		columnLineage = append(columnLineage, projection.ColumnLineage...)
		unresolved = append(unresolved, projection.UnresolvedReferences...)
	}
	return CompiledModelLineage{
		ColumnLineage:        columnLineage,
		UnresolvedReferences: unresolved,
		ProjectionCount:      len(selectItems),
	}
}

func splitCTEQueries(compiledSQL string) ([][2]string, string) {
	trimmed := strings.TrimLeft(compiledSQL, " \t\r\n")
	if len(trimmed) < 4 || strings.ToLower(trimmed[:4]) != "with" {
		return nil, compiledSQL
	}

	queries := make([][2]string, 0)
	index := 4
	for index < len(trimmed) {
		for index < len(trimmed) && isSpace(trimmed[index]) {
			index++
		}
		nameStart := index
		for index < len(trimmed) && (isAlphaNumeric(trimmed[index]) || trimmed[index] == '_') {
			index++
		}
		cteName := strings.TrimSpace(trimmed[nameStart:index])
		if cteName == "" {
			break
		}
		for index < len(trimmed) && isSpace(trimmed[index]) {
			index++
		}
		if index < len(trimmed) && trimmed[index] == '(' {
			index = consumeBalancedSegment(trimmed, index)
			for index < len(trimmed) && isSpace(trimmed[index]) {
				index++
			}
		}
		if index+2 > len(trimmed) || strings.ToLower(trimmed[index:index+2]) != "as" {
			break
		}
		index += 2
		for index < len(trimmed) && isSpace(trimmed[index]) {
			index++
		}
		if index >= len(trimmed) || trimmed[index] != '(' {
			break
		}
		queryStart := index + 1
		queryEnd := consumeBalancedSegment(trimmed, index) - 1
		queries = append(queries, [2]string{cteName, trimmed[queryStart:queryEnd]})
		index = queryEnd + 1
		for index < len(trimmed) && isSpace(trimmed[index]) {
			index++
		}
		if index < len(trimmed) && trimmed[index] == ',' {
			index++
			continue
		}
		return queries, trimmed[index:]
	}
	return nil, compiledSQL
}

func consumeBalancedSegment(text string, start int) int {
	depth := 0
	for index := start; index < len(text); index++ {
		switch text[index] {
		case '(':
			depth++
		case ')':
			depth--
			if depth == 0 {
				return index + 1
			}
		}
	}
	return len(text)
}

func extractSelectItems(sql string) []string {
	match := dbtSelectClauseRe.FindStringSubmatch(sql)
	if match == nil {
		return nil
	}
	selectClause := match[1]
	items := make([]string, 0)
	current := make([]rune, 0, len(selectClause))
	depth := 0
	inSingleQuote := false
	var prev rune
	for _, character := range selectClause {
		if character == '\'' && prev != '\\' {
			inSingleQuote = !inSingleQuote
		} else if !inSingleQuote {
			switch character {
			case '(':
				depth++
			case ')':
				if depth > 0 {
					depth--
				}
			case ',':
				if depth == 0 {
					item := strings.TrimSpace(string(current))
					if item != "" {
						items = append(items, item)
					}
					current = current[:0]
					prev = character
					continue
				}
			}
		}
		current = append(current, character)
		prev = character
	}
	if tail := strings.TrimSpace(string(current)); tail != "" {
		items = append(items, tail)
	}
	return items
}

func lineageForProjection(selectItem string, bindings map[string]*relationBinding, modelName string) CompiledModelLineage {
	matches := dbtAsAliasRe.FindStringSubmatch(strings.TrimSpace(selectItem))
	expression := strings.TrimSpace(selectItem)
	outputColumn := ""
	if matches != nil {
		expression = strings.TrimSpace(matches[1])
		outputColumn = strings.TrimSpace(matches[2])
	} else {
		outputColumn = implicitOutputColumn(expression)
	}

	if reason := expressionHonestyGapReason(expression); reason != "" {
		return CompiledModelLineage{UnresolvedReferences: []map[string]string{derivedExpressionGap(expression, modelName, reason)}, ProjectionCount: 1}
	}

	columnLineage := make([]ColumnLineage, 0)
	sourceColumns := make([]string, 0)
	unresolved := make([]map[string]string, 0)
	seenColumns := make(map[string]struct{})
	matchedIdentifiers := expressionIgnoredIdentifiers(expression)
	referenceExpression := referenceScanExpression(expression)

	for _, match := range qualifiedReferenceMatches(referenceExpression) {
		alias := match.Alias
		column := match.Column
		matchedIdentifiers[alias] = struct{}{}
		matchedIdentifiers[column] = struct{}{}
		columns, expanded, unresolvedRef := resolveQualifiedReference(bindings[alias], alias, column, modelName)
		if unresolvedRef != nil {
			unresolved = append(unresolved, unresolvedRef)
			continue
		}
		if len(expanded) > 0 {
			columnLineage = append(columnLineage, expanded...)
			continue
		}
		for _, sourceColumn := range columns {
			if _, ok := seenColumns[sourceColumn]; ok {
				continue
			}
			seenColumns[sourceColumn] = struct{}{}
			sourceColumns = append(sourceColumns, sourceColumn)
		}
	}

	for _, identifier := range unqualifiedIdentifiers(referenceExpression, matchedIdentifiers) {
		columns, unresolvedRef := resolveUnqualifiedReferenceColumns(identifier, bindings, modelName)
		if unresolvedRef != nil {
			unresolved = append(unresolved, unresolvedRef)
			continue
		}
		for _, sourceColumn := range columns {
			if _, ok := seenColumns[sourceColumn]; ok {
				continue
			}
			seenColumns[sourceColumn] = struct{}{}
			sourceColumns = append(sourceColumns, sourceColumn)
		}
	}

	if outputColumn == "" && len(sourceColumns) > 0 {
		outputColumn = sourceColumns[0][strings.LastIndex(sourceColumns[0], ".")+1:]
	}
	if len(sourceColumns) > 0 {
		if reason := expressionPartialReason(expression); reason != "" {
			unresolved = append(unresolved, derivedExpressionGap(expression, modelName, reason))
		}
	}
	if outputColumn != "" && len(sourceColumns) > 0 {
		metadata := transformMetadataForProjection(expression, bindings)
		columnLineage = append(columnLineage, ColumnLineage{
			OutputColumn:        strings.TrimSpace(outputColumn),
			SourceColumns:       sourceColumns,
			TransformKind:       metadata["transform_kind"],
			TransformExpression: metadata["transform_expression"],
		})
	}

	return CompiledModelLineage{
		ColumnLineage:        columnLineage,
		UnresolvedReferences: unresolved,
		ProjectionCount:      1,
	}
}

func referenceScanExpression(expression string) string {
	normalized := stripWrappingParentheses(strings.TrimSpace(expression))
	if inner, ok := unwrapTemplatedExpression(normalized); ok && !dbtBareIdentifierRe.MatchString(inner) && expressionPartialReason(inner) == "" {
		normalized = inner
	}
	matches := dbtQualifiedMacroCallRe.FindStringSubmatch(normalized)
	if matches == nil || (!isSupportedQualifiedMacroExpression(normalized) && !macroExpressionHasLineage(normalized)) {
		return normalized
	}
	return matches[2]
}

func unwrapTemplatedExpression(expression string) (string, bool) {
	normalized := strings.TrimSpace(expression)
	if !strings.HasPrefix(normalized, "{{") || !strings.HasSuffix(normalized, "}}") {
		return "", false
	}
	inner := strings.TrimSpace(strings.TrimSuffix(strings.TrimPrefix(normalized, "{{"), "}}"))
	if inner == "" {
		return "", false
	}
	return inner, true
}

func resolveQualifiedReference(binding *relationBinding, alias string, column string, modelName string) ([]string, []ColumnLineage, map[string]string) {
	if binding == nil {
		return nil, nil, derivedExpressionGap(alias+"."+column, modelName, "source_alias_not_resolved")
	}
	if column == "*" {
		if len(binding.columnNames) == 0 {
			return nil, nil, derivedExpressionGap(alias+".*", modelName, "wildcard_projection_not_supported")
		}
		expanded := make([]ColumnLineage, 0)
		if binding.assetName != "" {
			for _, expandedColumn := range binding.columnNames {
				expanded = append(expanded, ColumnLineage{
					OutputColumn:  expandedColumn,
					SourceColumns: []string{binding.assetName + "." + expandedColumn},
				})
			}
			return nil, expanded, nil
		}
		for _, expandedColumn := range binding.columnNames {
			item, ok := binding.columnLineage[expandedColumn]
			if !ok {
				continue
			}
			expanded = append(expanded, ColumnLineage{
				OutputColumn:  expandedColumn,
				SourceColumns: item.SourceColumns,
			})
		}
		return nil, expanded, nil
	}
	if binding.assetName != "" {
		return []string{binding.assetName + "." + column}, nil, nil
	}
	item, ok := binding.columnLineage[column]
	if !ok {
		return nil, nil, derivedExpressionGap(alias+"."+column, modelName, "cte_column_not_resolved")
	}
	return item.SourceColumns, nil, nil
}

func resolveUnqualifiedReferenceColumns(identifier string, bindings map[string]*relationBinding, modelName string) ([]string, map[string]string) {
	candidates := make(map[string][]string)
	for _, binding := range bindings {
		columns := bindingColumnsForIdentifier(binding, identifier)
		if len(columns) == 0 {
			continue
		}
		key := strings.Join(columns, "|")
		candidates[key] = columns
	}
	switch len(candidates) {
	case 0:
		return nil, derivedExpressionGap(identifier, modelName, "unqualified_column_reference_not_resolved")
	case 1:
		for _, columns := range candidates {
			return columns, nil
		}
	default:
		return nil, derivedExpressionGap(identifier, modelName, "unqualified_column_reference_ambiguous")
	}
	return nil, nil
}

func bindingColumnsForIdentifier(binding *relationBinding, identifier string) []string {
	if binding == nil {
		return nil
	}
	if binding.assetName != "" {
		for _, columnName := range binding.columnNames {
			if columnName == identifier {
				return []string{binding.assetName + "." + identifier}
			}
		}
		return nil
	}
	item, ok := binding.columnLineage[identifier]
	if !ok {
		return nil
	}
	return item.SourceColumns
}

func implicitOutputColumn(expression string) string {
	matches := dbtQualifiedReferenceScanRe.FindAllStringSubmatch(expression, -1)
	if len(matches) == 1 && matches[0][2] != "*" {
		return matches[0][2]
	}
	return ""
}

func transformMetadataForProjection(expression string, bindings map[string]*relationBinding) map[string]string {
	if metadata := expressionTransformMetadata(expression); metadata != nil {
		return metadata
	}
	return propagatedTransformMetadata(strings.TrimSpace(expression), bindings)
}

func propagatedTransformMetadata(expression string, bindings map[string]*relationBinding) map[string]string {
	normalized := strings.TrimSpace(expression)
	if dbtQualifiedReferenceRe.MatchString(normalized) {
		matches := qualifiedReferenceMatches(normalized)
		if len(matches) == 1 && matches[0].Column != "*" {
			return bindingTransformMetadata(bindings[matches[0].Alias], matches[0].Column)
		}
	}
	if !dbtBareIdentifierRe.MatchString(normalized) {
		return nil
	}
	candidates := make(map[string]map[string]string)
	for _, binding := range bindings {
		metadata := bindingTransformMetadata(binding, normalized)
		if metadata == nil {
			continue
		}
		key := metadata["transform_kind"] + "::" + metadata["transform_expression"]
		candidates[key] = metadata
	}
	if len(candidates) != 1 {
		return nil
	}
	for _, metadata := range candidates {
		return metadata
	}
	return nil
}

func bindingTransformMetadata(binding *relationBinding, columnName string) map[string]string {
	if binding == nil || binding.assetName != "" {
		return nil
	}
	item, ok := binding.columnLineage[columnName]
	if !ok || item.TransformKind == "" || item.TransformExpression == "" {
		return nil
	}
	return map[string]string{
		"transform_kind":       item.TransformKind,
		"transform_expression": item.TransformExpression,
	}
}

func isReservedAlias(alias string) bool {
	switch strings.ToLower(alias) {
	case "on", "where", "group", "order", "limit":
		return true
	default:
		return false
	}
}

func isSpace(value byte) bool {
	switch value {
	case ' ', '\t', '\n', '\r':
		return true
	default:
		return false
	}
}

func isAlphaNumeric(value byte) bool {
	return (value >= 'a' && value <= 'z') || (value >= 'A' && value <= 'Z') || (value >= '0' && value <= '9')
}
